nixosTest: Introduce IPv6

frederictobiasc opened the pull request:
## Description of changes

This contribution enables a working IPv6 setup by default. This works analog to the current automatic IPv4 setup.
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

frederictobiasc committed:
nixos/testing: Add ipv6 configuration

This contribution enables a working IPv6 setup by default. This works
analog to the current automatic IPv4 setup.

SuperSandro2000 commented on nixos/lib/testing/network.nix:
```diff
@@ -32,10 +32,19 @@ let
+          ];
+          ipv6.addresses = [
+            {
+              address = "2001:db8:${toString i.vlan}::${toString config.virtualisation.test.nodeNumber}";
```
We should probably use a range from fc00::/7 instead of the documentation range

rnhmjoj replied:
Yeah, that prefix should be limited to examples in documentation.
Using a ULA is better, but it need to be picked [randomly](https://en.wikipedia.org/wiki/Unique_local_address#Definition) to avoid interfering with existing tests.

frederictobiasc replied:
Thanks for your review!
It seems the default configuration prefers ULAs over IPv4 as well, so I don't have a strong opinion on that.
Do you think I should change it right now, or is there a chance we might get another opinion?

emilylange replied:
Using a GUA (here: the range reserved for documentation purposes) instead of a ULA has the benefit that IPv6 takes precedence over IPv4 when using `libc`'s `getaddrinfo`.

See [`gia.conf(5)`](https://man.archlinux.org/man/core/man-pages/gai.conf.5.en) and [`getaddrinfo(3)`](https://man.archlinux.org/man/getaddrinfo.3.en).

fpletz replied:
@emilylange Are you sure this is actually the case? As per [RFC 3484](https://www.ietf.org/rfc/rfc3484.txt) section 10.3:

> The default policy table gives IPv6 addresses higher precedence than IPv4 addresses.  This means that applications will use IPv6 in preference to IPv4 when the two are equally suitable.

Both should be equally suitable as per section 3.2 since private IPv4 ranges are treated as site-local - the same as the ULA prefix - and an address of the same prefix is assigned to an interface or there is a route that matches the destination.

I investigated this since the described behavior in general seemed a bit odd to me and didn't match the default configuration of `gai.conf`. I believe that IPv6 will always be preferred if a suitable IPv6 address/route is available since the IPv6 mapped IPv4 address prefix `::ffff:0:0/96` has the lowest precedence:

```
      Prefix        Precedence Label
      ::1/128               50     0
      ::/0                  40     1
      2002::/16             30     2
      ::/96                 20     3
      ::ffff:0:0/96         10     4
```

tfc commented:
@ofborgtest nat.firewall networking.scripted.link networking.scripted.virtual  networking.networkd.loopback  predictable-interface-names.unpredictableNetworkd  predictable-interface-names.unpredictable  networking.networkd.static networking.networkd.dhcpSimple networking.networkd.sit networking.networkd.bridge networking.networkd.privacy plasma5 systemd-networkd-ipv6-prefix-delegation networking.networkd.link networking.scripted.privacy  nat.firewall networking.scripted.link networking.scripted.virtual  networking.networkd.loopback  predictable-interface-names.unpredictableNetworkd  predictable-interface-names.unpredictable  networking.networkd.static networking.networkd.dhcpSimple networking.networkd.sit networking.networkd.bridge networking.networkd.privacy plasma5 systemd-networkd-ipv6-prefix-delegation networking.networkd.link networking.scripted.privacy

tfc commented:
@ofborg test nat.firewall networking.scripted.link networking.scripted.virtual  networking.networkd.loopback  predictable-interface-names.unpredictableNetworkd  predictable-interface-names.unpredictable  networking.networkd.static networking.networkd.dhcpSimple networking.networkd.sit networking.networkd.bridge networking.networkd.privacy plasma5 systemd-networkd-ipv6-prefix-delegation networking.networkd.link networking.scripted.privacy  nat.firewall networking.scripted.link networking.scripted.virtual  networking.networkd.loopback  predictable-interface-names.unpredictableNetworkd  predictable-interface-names.unpredictable  networking.networkd.static networking.networkd.dhcpSimple networking.networkd.sit networking.networkd.bridge networking.networkd.privacy plasma5 systemd-networkd-ipv6-prefix-delegation networking.networkd.link networking.scripted.privacy

tfc commented:
@ofborg test networking.networkd.macvlan networking.networkd.routes networking.networkd.dhcpOneIf networking.networkd.vlan  networking.scripted.routes networking.scripted.static  networking.scripted.sit networking.networkd.virtual networking.scripted.dhcpSimple networking.scripted.macvlan

tfc commented:
@ofborg test networking.networkd.bond networking.networkd.bridge networking.networkd.dhcpOneIf networking.networkd.dhcpSimple networking.networkd.link networking.networkd.loopback networking.networkd.macvlan networking.networkd.privacy

tfc commented:
@ofborg test networking.networkd.routes networking.networkd.sit networking.networkd.static networking.networkd.virtual networking.networkd.vlan networking.scripted.bond networking.scripted.bridge

tfc commented:
@ofborg test networking.scripted.bridge networking.scripted.dhcpOneIf networking.scripted.dhcpSimple networking.scripted.link networking.scripted.loopback networking.scripted.macvlan networking.scripted.privacy networking.scripted.routes networking.scripted.sit networking.scripted.static networking.scripted.virtual networking.scripted.vlan

tfc commented:
@ofborg test predictable-interface-names.predictable predictable-interface-names.predictableNetworkd predictable-interface-names.unpredictable predictable-interface-names.unpredictableNetworkd

frederictobiasc commented:
Hi, I checked the [newly failing jobs](https://hydra.nixos.org/eval/1806999#tabs-now-fail) of the related hydra run and pushed fixes to this branch.

Please check the following table for detailed investigation results:

| Investigation Result                | Job       | Finished at                                                     | Package/release name                               | System        |
| ----------------------------------- | --------- | --------------------------------------------------------------- | -------------------------------------------------- | ------------- |
| fails on aarch64 only, cannot check | 263166476 | tests.fcitx5.aarch64-linux                                      | 2d ago vm-test-run-fcitx5                          | aarch64-linux |
| fails on aarch64 only, cannot check | 263166682 | tests.ferm.aarch64-linux                                        | 2d ago vm-test-run-ferm                            | aarch64-linux |
| flaky, mitigated                    | 263165379 | tests.firewall-nftables.aarch64-linux                           | 2d ago vm-test-run-firewall-nftables               | aarch64-linux |
| flaky, mitigated                    | 263166144 | tests.firewall-nftables.x86_64-linux                            | 2d ago vm-test-run-firewall-nftables               | x86_64-linux  |
| not reproducible                    | 263167706 | tests.home-assistant.x86_64-linux                               | 2d ago vm-test-run-home-assistant                  | x86_64-linux  |
| fails on aarch64 only, cannot check | 263167984 | tests.incus-lts.container-systemd-init.aarch64-linux            | 2d ago vm-test-run-container-systemd-init          | aarch64-linux |
| fails on aarch64 only, cannot check | 263164937 | tests.installed-tests.ibus.aarch64-linux                        | 2d ago vm-test-run-ibus-1.5.30                     | aarch64-linux |
| flaky                               | 263167484 | tests.installer-systemd-stage-1.clevisLuksFallback.x86_64-linux | 2d ago vm-test-run-installer-clevis-luks-fallback  | x86_64-linux  |
| fails on aarch64 only, cannot check | 263165543 | tests.iodine.aarch64-linux                                      | 2d ago vm-test-run-iodine                          | aarch64-linux |
| fixed                               | 263169749 | tests.ipv6.aarch64-linux                                        | 2d ago vm-test-run-ipv6                            | aarch64-linux |
| fixed                               | 263165469 | tests.ipv6.x86_64-linux                                         | 2d ago vm-test-run-ipv6                            | x86_64-linux  |
| fixed                               | 263166506 | tests.iscsi-root.x86_64-linux                                   | 2d ago vm-test-run-iscsi                           | x86_64-linux  |
| fixed                               | 263169513 | tests.jool.nat64.aarch64-linux                                  | 2d ago vm-test-run-jool-nat64                      | aarch64-linux |
| fixed                               | 263166388 | tests.jool.nat64.x86_64-linux                                   | 2d ago vm-test-run-jool-nat64                      | x86_64-linux  |
| fails on aarch64 only, cannot check | 263165238 | tests.keymap.colemak.aarch64-linux                              | 2d ago vm-test-run-keymap-colemak                  | aarch64-linux |
| flaky                               | 263165089 | tests.librenms.x86_64-linux                                     | 2d ago vm-test-run-librenms                        | x86_64-linux  |
| not reproducible                    | 263167839 | tests.lightdm.x86_64-linux                                      | 2d ago vm-test-run-lightdm                         | x86_64-linux  |
| unrelated                           | 263168694 | tests.mariadb-galera.mariadb_1011.aarch64-linux                 | 2d ago vm-test-run-mariadb_1011-galera-mariabackup | aarch64-linux |
| unrelated                           | 263166624 | tests.mariadb-galera.mariadb_1011.x86_64-linux                  | 2d ago vm-test-run-mariadb_1011-galera-mariabackup | x86_64-linux  |
| unrelated                           | 263170168 | tests.mariadb-galera.mariadb_105.aarch64-linux                  | 2d ago vm-test-run-mariadb_105-galera-mariabackup  | aarch64-linux |
| unrelated                           | 263167541 | tests.mariadb-galera.mariadb_105.x86_64-linux                   | 2d ago vm-test-run-mariadb_105-galera-mariabackup  | x86_64-linux  |
| unrelated                           | 263166788 | tests.mariadb-galera.mariadb_106.aarch64-linux                  | 2d ago vm-test-run-mariadb_106-galera-mariabackup  | aarch64-linux |
| unrelated                           | 263167782 | tests.mariadb-galera.mariadb_106.x86_64-linux                   | 2d ago vm-test-run-mariadb_106-galera-mariabackup  | x86_64-linux  |
| unrelated                           | 263169668 | tests.mariadb-galera.mariadb_110.aarch64-linux                  | 2d ago vm-test-run-mariadb_110-galera-mariabackup  | aarch64-linux |
| unrelated                           | 263164961 | tests.mariadb-galera.mariadb_110.x86_64-linux                   | 2d ago vm-test-run-mariadb_110-galera-mariabackup  | x86_64-linux  |
| fixed                               | 263167681 | tests.mediatomb.aarch64-linux                                   | 2d ago vm-test-run-medi12atomb                     | aarch64-linux |
| fixed                               | 263168717 | tests.mediatomb.x86_64-linux                                    | 2d ago vm-test-run-mediatomb                       | x86_64-linux  |
| not reproducible                    | 263166740 | tests.musescore.x86_64-linux                                    | 2d ago vm-test-run-musescore                       | x86_64-linux  |
| not reproducible                    | 263168896 | tests.nat.firewall.aarch64-linux                                | 2d ago vm-test-run-natWithFirewall                 | aarch64-linux |
| not reproducible                    | 263169751 | tests.nat.firewall.x86_64-linux                                 | 2d ago vm-test-run-natWithFirewall                 | x86_64-linux  |
| not reproducible                    | 263165365 | tests.nat.nftables.firewall.aarch64-linux                       | 2d ago vm-test-run-natNftablesWithFirewall         | aarch64-linux |
| not reproducible                    | 263169501 | tests.nat.nftables.firewall.x86_64-linux                        | 2d ago vm-test-run-natNftablesWithFirewall         | x86_64-linux  |
| not reproducible                    | 263168752 | tests.nat.nftables.standalone.aarch64-linux                     | 2d ago vm-test-run-natNftablesStandalone           | aarch64-linux |
| not reproducible                    | 263169024 | tests.nat.nftables.standalone.x86_64-linux                      | 2d ago vm-test-run-natNftablesStandalone           | x86_64-linux  |
| fixed                               | 263170439 | tests.nat.standalone.aarch64-linux                              | 2d ago vm-test-run-natStandalone                   | aarch64-linux |
| fixed                               | 263168417 | tests.nat.standalone.x86_64-linux                               | 2d ago vm-test-run-natStandalone                   | x86_64-linux  |
| fixed                               | 263166245 | tests.nvmetcfg.aarch64-linux                                    | 2d ago vm-test-run-nvmetcfg                        | aarch64-linux |
| fixed                               | 263166020 | tests.nvmetcfg.x86_64-linux                                     | 2d ago vm-test-run-nvmetcfg                        | x86_64-linux  |
| not reproducible                    | 263166781 | tests.osquery.x86_64-linux                                      | 2d ago vm-test-run-osquery                         | x86_64-linux  |
| not reproducible                    | 263166511 | tests.qtile.x86_64-linux                                        | 2d ago vm-test-run-qtile                           | x86_64-linux  |
| not reproducible                    | 263168193 | tests.slurm.x86_64-linux                                        | 2d ago vm-test-run-slurm                           | x86_64-linux  |
| fixed, unrelated                    | 263170064 | tests.step-ca.x86_64-linux                                      | 2d ago vm-test-run-step-ca                         | x86_64-linux  |
| fails on aarch64 only, cannot check | 263170315 | tests.systemd-initrd-vconsole.aarch64-linux                     | 2d ago vm-test-run-systemd-initrd-vconsole         | aarch64-linux |
| unrelated                           | 263166286 | tests.tmate-ssh-server.aarch64-linux                            | 2d ago vm-test-run-tmate-ssh-server                | aarch64-linux |
| unrelated                           | 263169688 | tests.tmate-ssh-server.x86_64-linux                             | 2d ago vm-test-run-tmate-ssh-server                | x86_64-linux  |
| unrelated                           | 263166243 | tests.tuxguitar.x86_64-linux                                    | 2d ago vm-test-run-tuxguitar                       | x86_64-linux  |
| not reproducible                    | 263165600 | tests.uptermd.x86_64-linux                                      | 2d ago vm-test-run-uptermd                         | x86_64-linux  |
| fixed                               | 263165627 | tests.vaultwarden.sqlite.x86_64-linux                           | 2d ago vm-test-run-vaultwarden-sqlite              | x86_64-linux  |
| fixed                               | 263169221 | tests.vector.dnstap.x86_64-linux                                | 2d ago vm-test-run-vector-dnstap                   | x86_64-linux  |
| fixed                               | 263168890 | tests.vector.dnstap.aarch64-linux                               | 2d ago vm-test-run-vector-dnstap                   | aarch64-linux |
| not reproducible                    | 263167185 | tests.vsftpd.x86_64-linux                                       | 2d ago vm-test-run-vsftpd                          | x86_64-linux  |

## Explanation of `Investigation Result`

- unrelated: Fails, but seems to be unrelated to IPv6 the implementation of IPv6
- not reproducible: Works on my machine, probably a flaky test.
- fixed: There is a commit fixing the test
- fixed, unrelated: Like fixed, but the failing test is unrelated to the implementation of IPv6
- flaky: Sometimes, the test succeeds, sometimes it fails.
- flaky, mitigated: Like flaky, but there is a commit mitigating commit.

SuperSandro2000 commented on nixos/tests/firewall.nix:
```diff
@@ -60,7 +60,10 @@ import ./make-test-python.nix ( { pkgs, nftables, ... } : {
-    # Check whether activation of a new configuration reloads the firewall.
+    # This sleep seems to mitigate some race condition that prevents the firewall being
+    # reloaded/started as part of the activation of the new config.
+    walled.sleep(1)
```
Can we somehow wait or check something? one second might not be enough for the hydra builders

frederictobiasc replied:
I have no idea what to wait for. The [stop_job](https://github.com/NixOS/nixpkgs/blob/edbebf5ce4cba33e777eb038aa9b95f947bef518/nixos/tests/firewall.nix#L60) blocks until the unit is stopped. Even the subsequent curl request on the previously walled resource finishes.
But this means in fact that the D-Bus response to `switch-to-configuration.pl`'s `get_active_units()` contains an outdated state.

Since it's not related to this PR, I reverted the change, and I'd like to leave it as is.

frederictobiasc commented:
> Can you please drop the formatting changes from this PR? I agree with them but it makes this PR rather big and hard to review and I kinda got lost in them.

Done

tfc commented:
Thank you for your efforts!

vcunat commented:
This merge blocks the `nixos-unstable` channel now.  A test started to hang reproducibly ([example log](https://hydra.nixos.org/build/266579247/nixlog/3/tail)), locally e.g.
```bash
nix build -f. nixosTests.systemd-networkd-ipv6-prefix-delegation
```

frederictobiasc commented:
Hi @vcunat,

I'm already preparing a fix. Since I only checked the [newly failing jobs](https://hydra.nixos.org/eval/1806999#tabs-now-fail) of the related hydra run, I did not fix this test.

vcunat commented:
The comparison base can be selected arbitrarily, BTW.

frederictobiasc commented:
https://github.com/NixOS/nixpkgs/pull/328375
